File: go-example/scripts/import/import.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
)

//
// Exit Codes
//

const (
	EXIT_CODE_VALIDATION_ERROR     = 1
	EXIT_CODE_TRANSFORMATION_ERROR = 2
	EXIT_CODE_UNEXPECTED_ERROR     = 3
)

//
// Utilities
//

func logMessage(message string) {
	_, _ = fmt.Fprintln(os.Stderr, message)
}

func printWarning(message string) {
	fmt.Println(message)
}

func verifyDir(path string) {
	info, err := os.Stat(path)

	if err != nil {
		if os.IsNotExist(err) {
			logMessage("path " + path + " does not exist")
			os.Exit(EXIT_CODE_UNEXPECTED_ERROR)
		} else {
			logMessage("unexpected error: " + err.Error())
			os.Exit(EXIT_CODE_UNEXPECTED_ERROR)
		}
	}

	if !info.IsDir() {
		logMessage("path " + path + " is not a directory")
		os.Exit(EXIT_CODE_UNEXPECTED_ERROR)
	}
}

func copyFile(from, to string) {
	read, err := os.Open(from)
	if err != nil {
		logMessage("unexpected error: " + err.Error())
		os.Exit(EXIT_CODE_UNEXPECTED_ERROR)
	}
	defer read.Close()

	write, err := os.Create(to)
	if err != nil {
		logMessage("unexpected error: " + err.Error())
		os.Exit(EXIT_CODE_UNEXPECTED_ERROR)
	}
	defer write.Close()

	write.ReadFrom(read)
}

//
// Parameter Handling
//

func main() {
	if len(os.Args) < 3 {
		logMessage("called with too few arguments, callers must provide an input and output directory")
		os.Exit(EXIT_CODE_UNEXPECTED_ERROR)
	}

	inputDirectory := os.Args[1]
	outputDirectory := os.Args[2]

	verifyDir(inputDirectory)
	verifyDir(outputDirectory)

	runImport(inputDirectory, outputDirectory)
}

//
// Script Body
//

func runImport(inputDir, outputDir string) {
	logMessage("Here's a log message that will be printed to STDERR")

	files, err := os.ReadDir(inputDir)
	if err != nil {
		logMessage("unexpected error: " + err.Error())
		os.Exit(EXIT_CODE_UNEXPECTED_ERROR)
	}

	for _, file := range files {
		logMessage("Copying " + file.Name() + " to " + outputDir)
		printWarning("Here's a validation warning about " + file.Name() + " that will be printed to STDOUT")
		copyFile(filepath.Join(inputDir, file.Name()), filepath.Join(outputDir, file.Name()))
	}
}

File: python-example/scripts/import.py
#!/usr/bin/env python3

import os
import shutil
import sys

#
# Exit Codes
#

EXIT_CODE_VALIDATION_ERROR = 1
EXIT_CODE_TRANSFORMATION_ERROR = 2
EXIT_CODE_UNEXPECTED_ERROR = 3

#
# Utilities
#


def log_message(*args):
    print(*args, file=sys.stderr)


def print_warning(msg):
    print(msg)


def verify_dir(path):
    if path == "" or path is None:
        log_message("required directory parameter was blank or absent")
        sys.exit(EXIT_CODE_UNEXPECTED_ERROR)

    if not os.path.isdir(path):
        log_message(f"directory {path} does not exist or is not a directory")
        sys.exit(EXIT_CODE_UNEXPECTED_ERROR)

#
# Parameter Handling
#


if len(sys.argv) < 3:
    log_message("called with too few arguments, callers must provide an input and output directory")
    sys.exit(EXIT_CODE_UNEXPECTED_ERROR)

verify_dir(sys.argv[1])
verify_dir(sys.argv[2])

INPUT_DIR = sys.argv[1]
OUTPUT_DIR = sys.argv[2]

#
# Script Body
#

log_message("Here's a log message that will be printed to STDERR")

for file in os.listdir(INPUT_DIR):
    log_message(f"copying {file} to {OUTPUT_DIR}")
    print_warning(f"Here's a validation warning about {file} that will be printed to STDOUT")
    shutil.copy2(os.path.join(INPUT_DIR, file), os.path.join(OUTPUT_DIR, file))
